lasy/propagators/collins_sfft_propagator.py

```diff
@@ -1,9 +1,283 @@
+
+        self.dim = dim
+        self.omega0 = omega0 if omega0 is not None else self.omega0
+        self.abcd = abcd  # optical ray matrix
+
+    def add_vacuum(self, distance):
+        vacuum = np.array([[1, distance], [0, 1]])
+        self.abcd = np.matmul(vacuum, self.abcd)
+        return self.abcd
+
+    def add_lens(self, focal_length):
+        lens = np.array([[1, 0], [-1.0 / focal_length, 1]])
+        self.abcd = np.matmul(lens, self.abcd)
+        return self.abcd
+
+    def reset_matrix(self):
+        self.abcd = np.array([[1, 0], [0, 1]])
+        return self.abcd
+
+    def add_output_grid(self, dim, grid_in):
+        """
+        Function to calculate the output grids
+        for a focusing / defocusing beam
+
+        Parameters
+        ----------
+        dim : string
+            Dimensionality of the array. Options are:
+            - ``'xyt'``: Laser pulse represented on a 3D Cartesian grid.
+            - ``'rt'`` : Laser pulse represented on a 2D cylindrical grid.
+
+        grid_in : meshgrid (in meter)
+            2D meshgrid for the input coordinates
+
+        """
+        if self.dim == "rt":
+            print(
+                "'rt' geometry not yet supported by CollinsSFFTPropagator, skipping grid calculation"
+            )
+            grid_out = deepcopy(grid_in)  # Make a copy of the input grid
+
+        else:  # self.dim == "xyt"
+            grid_out = deepcopy(grid_in)  # Make a copy of the input grid
+
+            try:  # Get the elements of the optical matrix
+                A = self.abcd[0][0]
+                B = self.abcd[0][1]
+                C = self.abcd[1][0]
+                D = self.abcd[1][1]
+                print(
+                    "Determinant of optical matrix: ", A * D - B * C
+                )  # Check determinant = 1
+            except:
+                print("Missing the ray matrix for the optical system.")
+
+            x = grid_in.axes[0]
+            y = grid_in.axes[1]
+            L0_width = np.abs(x[-1] - x[0])
+            N_points = len(x)
+
+            lambda0 = 2.0 * np.pi * c / self.omega0
+            k0 = self.omega0 / c
+
+            w0 = get_w0(grid_in, self.dim)  # Calculate input spot size
+            z_R = np.pi * w0**2 / lambda0  # Calculate input Rayleigh range
+
+            q1 = q(0, 0, z_R)
+            z_02 = -np.real((A * q1 + B) / (C * q1 + D))  # Calculate waist position
+            z_R2 = -np.imag(
+                (A * q1 + B) / (C * q1 + D)
+            )  # Calculate output Rayleigh range
+            f0 = np.sqrt(k0 * w0**2 / 2.0 * z_R2)  # Calculate effective focal length
+            assert f0 < 5, (
+                "CollinsSFFTPropagator is for focusing geometries, please specify a lens."
+            )
+
+            r0_step = L0_width / N_points  # Note: D gridpoints means D-1 intervals
+
+            x = fftshift(fftfreq(N_points, r0_step) * lambda0 * f0)
+            y = fftshift(fftfreq(N_points, r0_step) * lambda0 * f0)
```

Comment:
Do I understand correctly that, here, you need the frequency axis but without doing any FFT? The current fft wrapper implementation does not allow for this, but we could add a function e.g. `fft_axis` or so. Would you then be able to use it?